source/slang/slang-ir-specialize.cpp

```diff
@@ -226,7 +227,9 @@ struct SpecializationContext
         UInt argCount = specializeInst->getArgCount();
         for (UInt ii = 0; ii < argCount; ++ii)
         {
-            key.vals.add(specializeInst->getArg(ii));
+            IRInst* genericArg = specializeInst->getArg(ii);
```

Comment:
It would be better if we can also need to specialize the generic using the folded args.

Reply:
I will find a better place to do the folding.

Reply:
I was trying to find a better place to do the folding; preferably earlier location than here.
But it seems like this is the best place to fold.
Because here is the first place that branches to either find one or generate one.

I look back on your comment and if I take "also" as a keyword, I can find another location to do the folding at a location earlier than "BEFORE-SPECIALIZE" point.
The last part of a function, "cloneGlobalValueWithCodeCommon()", is where generic functions are copied.
I am not sure why copies are happening, but it seems to make copies two copies for the given test case below.

Maybe I need to look for a place where the initial one gets built up and do the folding there.

I can use some help based on what I have learned so far. @csyonghe

Reply:
Sorry about the confusion.

This is the best place to try folding.

My previous comment is that it is better if we can refactor specializeGenericImpl function called down below to take in the newly folded args and use them to perform the actual specialization, so we generate more compact specialized code.